Adjust engineer filter key

ChessBerry opened the pull request:
This PR adds two changes to the reintroduced SelectHighestEngineerAndAssist hotkey.

1. It treats T3 engies and SACUs the same for the purpose of selecting the highest tech level constructors. This is preferable as the tech level a unit can produce is more important than what the unit actually is, so with the new behavior where we select not one but all equivalent engies, we want to also treat t3 engies and sacus the same.

2. If you have multiple different factions in your selection, it now selects the faction's engies of which you have the most of for the highest tech level. For example, if you have 4 t3 uef engies and 2 t3 aeon engies, it will select the 4 t3 uef engies and have the 2 aeon engies assist those. Works for purely t1 engies too.
This mostly matters in fullshare team games when you get the base of an ally but in those situations such a hotkey can be a lifesaver.

ChessBerry commented:
Note that I'm guessing that there is a much smarter way to count the number of units of each faction than using the utility function CalcTableLength I have written above, but a quick google search just told me that lua tables by default seemingly don't have a length parameter.
Still, I'm guessing there already is a better function in the FA repo that does this that I'm not aware about

Garanas commented:
Is this written with ChatGPT?

ChessBerry commented:
> Is this written with ChatGPT?

What? xd

Tbf I have been using ChatGPT quite a bit over the last months so maybe it has rubbed off on me. 

But what if if this too is written by ChatGPT and I just didn't notice? What if everything I write is actually ChatGPT?? WHAT IF I AM JUST SIMULATED BY CHATGPT??? AM I REAL?!?

Garanas commented:
And that is how I accidentally triggered an existential crisis 😄 

I'll rewrite it

HUSSAR-mtrela commented on lua/keymap/misckeyactions.lua:
```diff
@@ -546,26 +546,55 @@ RestoreCameraPosition = function ()
+    local faction_nr = 0
+    local nr_of_units = 0
+    for i, u in units_factions do
+        nr_of_units_for_that_faction = CalcTableLength(u)
```
replace CalcTableLength with global function called: table.getsize

HUSSAR-mtrela commented on lua/keymap/misckeyactions.lua:
```diff
@@ -546,26 +546,55 @@ RestoreCameraPosition = function ()
     camera:MoveTo(settings.Focus, { settings.Heading, settings.Pitch, 0}, settings.Zoom, 0)
 end
 
+function CalcTableLength(tbl)
```
remove CalcTableLength function since it is the same as table.getsize function

HUSSAR-mtrela commented on lua/keymap/misckeyactions.lua:
```diff
@@ -546,26 +546,55 @@ RestoreCameraPosition = function ()
-        elseif next(tech3) then
-            SimCallback({Func= 'SelectHighestEngineerAndAssist', Args = { TargetId = tech3[1]:GetEntityId() }}, true)
-            SelectUnits(tech3)
+        local tech3_and_sACUs = EntityCategoryFilterDown((categories.SUBCOMMANDER + categories.TECH3) - categories.COMMAND, selection)
```
remove "_" in name of variable and use camelCase for variable

HUSSAR-mtrela commented on lua/keymap/misckeyactions.lua:
```diff
@@ -546,26 +546,55 @@ RestoreCameraPosition = function ()
-            SelectUnits(tech3)
+        local tech3_and_sACUs = EntityCategoryFilterDown((categories.SUBCOMMANDER + categories.TECH3) - categories.COMMAND, selection)
+
+        local highest_tech_engies_and_sacus_of_majority_faction = nil
```
remove "_" in name of variable and use camelCase for variable

HUSSAR-mtrela commented on lua/keymap/misckeyactions.lua:
```diff
@@ -546,26 +546,55 @@ RestoreCameraPosition = function ()
+end
+
+function GetMajorityFaction(units)
+    local faction_categories = {categories.UEF, categories.CYBRAN, categories.AEON, categories.SERAPHIM}
```
remove "_" in name of variable and use camelCase for variable

HUSSAR-mtrela commented on lua/keymap/misckeyactions.lua:
```diff
@@ -546,26 +546,55 @@ RestoreCameraPosition = function ()
+function GetMajorityFaction(units)
+    local faction_categories = {categories.UEF, categories.CYBRAN, categories.AEON, categories.SERAPHIM}
+
+    local units_factions = {}
```
remove "_" in name of variable and use camelCase for variable

ChessBerry commented:
Apologies, I completely forgot this PR existed..

I merged it with the current develop branch, implemented the suggestions by @HUSSAR-mtrela and fixed a bug where engies accidentally assisted army units sometimes.

I think it works as intended now and will try my hardest to not forget the existence of this PR again for a year :P

ChessBerry commented:
Should maybe ping @Garanas here too

lL1l1 commented on lua/ui/game/hotkeys/filter-engineers.lua:
```diff
@@ -21,27 +21,51 @@
+            highestTechEngiesAndSacusOfMajorityFaction = GetMajorityFaction(tech2Engineers)
+        elseif next(tech1Engineers) then
+            highestTechEngiesAndSacusOfMajorityFaction = GetMajorityFaction(tech1Engineers)
         else
             -- do nothing
```
```suggestion
```

lL1l1 commented on lua/ui/game/hotkeys/filter-engineers.lua:
```diff
@@ -21,27 +21,51 @@
-        elseif next(tech2) then
-            SimCallback({ Func = 'SelectHighestEngineerAndAssist', Args = { TargetId = tech2[1]:GetEntityId() } }, true)
-            SelectUnits(tech2)
+        local tech3EngineersAndSACUs = EntityCategoryFilterDown((categories.ENGINEER * categories.TECH3 + categories.SUBCOMMANDER) - categories.COMMAND, selection)
+        local tech2Engineers = EntityCategoryFilterDown(categories.ENGINEER * categories.TECH2 - categories.COMMAND, selection)
+        local tech1Engineers = EntityCategoryFilterDown(categories.ENGINEER * categories.TECH1 - categories.COMMAND, selection)
```
[Suggested change] Cache the category calculation in a local variable at the top of the file. Like here:
https://github.com/FAForever/fa/blob/4eed87d6df67e9691b259eefb7326c7b4227c944/lua/shield.lua#L98-L99

ChessBerry replied:
Implemented

ChessBerry commented on lua/ui/game/hotkeys/filter-engineers.lua:
```diff
@@ -21,27 +21,51 @@
 
+--- Filters your selection to the units of the majority faction in your selection.
+function GetMajorityFaction(units)
+    local factionCategories = {categories.UEF, categories.CYBRAN, categories.AEON, categories.SERAPHIM}
```
Implemented

ChessBerry commented on lua/ui/game/hotkeys/filter-engineers.lua:
```diff
@@ -21,27 +21,51 @@
+        end
+    end
+
+    return factionsOfUnits[factionNr]
```
Implemented

ChessBerry commented on lua/ui/game/hotkeys/filter-engineers.lua:
```diff
@@ -21,27 +21,51 @@
+    return factionsOfUnits[factionNr]
+end
+
 --- Filters your selection to the highest tech of engineers. All other engineers assist one of those engineers.
```
I have updated the hotkey description in `keydescriptions.lua` but I haven't touched `keydescriptions.lua` because seemingly none of the advanced order hotkeys have an entry there yet.

ChessBerry commented on lua/ui/game/hotkeys/filter-engineers.lua:
```diff
@@ -21,27 +21,51 @@
-            SelectUnits(tech2)
+        local tech3EngineersAndSACUs = EntityCategoryFilterDown((categories.ENGINEER * categories.TECH3 + categories.SUBCOMMANDER) - categories.COMMAND, selection)
+        local tech2Engineers = EntityCategoryFilterDown(categories.ENGINEER * categories.TECH2 - categories.COMMAND, selection)
+        local tech1Engineers = EntityCategoryFilterDown(categories.ENGINEER * categories.TECH1 - categories.COMMAND, selection)
```
Implemented the sparky filtering with a priority between t1 and t2 engies.

I'll leave the fully general buildable category filtering for a future PR.

ChessBerry commented on lua/ui/game/hotkeys/filter-engineers.lua:
```diff
@@ -80,7 +83,7 @@ function SelectHighestEngineerAndAssist()
         end
 
-        if highestTechEngiesAndSacusOfMajorityFaction then
+        if highestTechEngiesAndSacusOfMajorityFaction and table.getn(selection) ~= table.getn(highestTechEngiesAndSacusOfMajorityFaction) then
```
I think it's preferable to keep the callback when units found equals selection as the callback cancels queued orders and this behavior should imo be consistent whenever you use the hotkey.
